Sources/Datadog/DatadogCore/Context/DatadogContextProvider.swift

```diff
@@ -149,8 +146,16 @@ internal final class DatadogContextProvider {
     ///   - reader: The value reader.
     ///   - keyPath: A context's key path that supports reading from and writing to the resulting value.
     func assign<Reader>(reader: Reader, to keyPath: WritableKeyPath<DatadogContext, Reader.Value>) where Reader: ContextValueReader {
-        queue.async(flags: .barrier) {
+        queue.async {
             self.reader.append(reader: reader, receiver: keyPath)
         }
     }
+
+#if DD_SDK_COMPILED_FOR_TESTING
+    func replace(context newContext: DatadogContext) {
+        queue.async {
+            self.context = newContext
+        }
+    }
+#endif
```

Comment:
I'm very much in favour of avoiding this 👍, but it will require even bigger diff in this PR.

Rationale of this hack is **solely** to facilitate the switch from previous `DatadogCoreMock` to new `DatadogCoreProxy` in tests. The way we used `DatadogCoreMock` was this:
```swift
private var core: DatadogCoreMock!

func testSomething() {
   core.context = .mockWith(
      // ... configure the mock context
   )

   // use the core in test
}
```
so I had to add analogous behaviour of changing `.context` lazily on `DatadogCoreProxy`:
```swift
// DatadogCoreProxy

var context: DatadogContext {
   didSet {
      let core = self.proxy as! DatadogCore
      core.contextProvider.replace(context: context)
   }
}
```

IMO this is a flaw in our testing convention: the `.context` should be managed only by the core, without allowing anyone else (including `DatadogCoreProxy`) to modify it arbitrarily. Migrating our tests to this following model will fulfil this idea:
```swift
func testSomething() {
   let core = DatadogCoreProxy(
      context = .mockWith(
         // ... configure the mock context
      )
   )
   defer { core.flushAndTearDown() }

   // use the core in test
}
```
The price will be to refactor all places (roughly ~100 tests). If we agree on this model, I'm OK with doing that in separate PR to isolate changes. WDYT @maxep @maciejburda ?

Reply:
I like the idea. It's better to do this now than never.

Reply:
I made an attempt to resolve this and use new model in tests, but gave up as it requires slightly more refactoring in tests, that I don't want to conduct right now.

I got blocked on this tests (there are 2 like this in each feature: Logging, Tracing and RUM):
https://github.com/DataDog/dd-sdk-ios/blob/fcf5daa3155d3e29617a193d0fba3987b6569e01/Tests/DatadogTests/Datadog/LoggerTests.swift#L325-L331

What happens here is:
* simulate _carrier info_ / _network info_ change
* send Log | Span | RUM event
* assert the event has proper _carrier info_ / _network info_ set

Speaking V1: it made sense to test this behaviour in each Feature separately.
Speaking V2: this certain behaviour should be tested in `DatadogCoreTests`, only once.

Because we don't yet have `DatadogCoreTests` defined, I'm leaving this for now. We will revisit this later in migration.